pkg/operator/util/apply.go

```diff
@@ -21,3 +40,79 @@ func ApplyAsset(h *resource.Helper, assetPath string, hLog log.FieldLogger) erro
+		return err
+	}
+	assetLog.Info("applying asset with GC")
+	result, err := ApplyRuntimeObjectWithGC(h, runtimeObj, hc)
+	if err != nil {
+		assetLog.WithError(err).Error("error applying asset")
+		return err
+	}
+	assetLog.Infof("asset applied successfully: %v", result)
+	return nil
+}
+
+// ApplyAssetWithNSOverrideAndGC loads the given asset, overrides the namespace, adds an owner reference to
+// HiveConfig for uninstall, and applies it to the cluster.
+func ApplyAssetWithNSOverrideAndGC(h *resource.Helper, assetPath, namespaceOverride string, hiveConfig *hivev1.HiveConfig) error {
+	requiredObj, _, err := coreDeserializer.Decode(assets.MustAsset(assetPath), nil, nil)
+	if err != nil {
+		return errors.Wrapf(err, "unable to decode asset: %s", assetPath)
+	}
+	obj, _ := meta.Accessor(requiredObj)
+	obj.SetNamespace(namespaceOverride)
+	_, err = ApplyRuntimeObjectWithGC(h, requiredObj, hiveConfig)
+	if err != nil {
+		return errors.Wrapf(err, "unable to apply asset: %s", assetPath)
+	}
+	return nil
+}
+
+// ApplyClusterRoleBindingAssetWithSubjectNSOverrideAndGC loads the given asset, overrides the namespace on the subject,
+// adds an owner reference to HiveConfig for uninstall, and applies it to the cluster.
+func ApplyClusterRoleBindingAssetWithSubjectNSOverrideAndGC(h *resource.Helper, roleBindingAssetPath, namespaceOverride string, hiveConfig *hivev1.HiveConfig) error {
+
+	rb := resourceread.ReadClusterRoleBindingV1OrDie(assets.MustAsset(roleBindingAssetPath))
+	for i := range rb.Subjects {
+		if rb.Subjects[i].Kind == "ServiceAccount" || rb.Subjects[i].Namespace != "" {
+			rb.Subjects[i].Namespace = namespaceOverride
+		}
+	}
+	_, err := ApplyRuntimeObjectWithGC(h, rb, hiveConfig)
+	if err != nil {
+		return errors.Wrapf(err, "unable to apply asset: %s", roleBindingAssetPath)
+	}
+	return nil
+}
+
+// ApplyRuntimeObjectWithGC adds an OwnerReference to the HiveConfig on the runtime object, and applies it to the cluster.
+func ApplyRuntimeObjectWithGC(h *resource.Helper, runtimeObj runtime.Object, hc *hivev1.HiveConfig) (resource.ApplyResult, error) {
+	obj, err := meta.Accessor(runtimeObj)
+	if err != nil {
+		return resource.UnknownApplyResult, err
+	}
+	ownerRef := v1.OwnerReference{
+		APIVersion:         hc.APIVersion,
+		Kind:               hc.Kind,
+		Name:               hc.Name,
+		UID:                hc.UID,
+		BlockOwnerDeletion: pointer.BoolPtr(true),
+	}
+	// This assumes we have full control of owner references for these resources the operator creates.
+	obj.SetOwnerReferences([]v1.OwnerReference{ownerRef})
+	return h.ApplyRuntimeObject(runtimeObj, scheme.Scheme)
+}
+
+func readRuntimeObject(assetPath string) (runtime.Object, error) {
+	codec := serializer.NewCodecFactory(scheme.Scheme)
+	// Additional SchemeGroupVersions may need to be added here if the operator needs to start deploying new types.
+	return runtime.Decode(codec.UniversalDecoder(rbacv1.SchemeGroupVersion, orbacv1.GroupVersion), assets.MustAsset(assetPath))
```

Comment:
You want `rbacv1` and `orbacv1` added to the scheme not added as preferred versions to use. If they are not in the scheme, then the decoder will not know how to decode them.

Reply:
Also, for what it is worth, you should not have to decode to an structured type to add the owner reference. You should be able to decode to an unstructured type.

Reply:
I am very confused, my first attempt at fixing this error (which I got for the Kube type) was to add to the scheme in cmd/operator/main.go, but this did not work. I then started passing the scheme all around from manager down to this function to make sure the right one was in use, but that also didn't work. Adding as an argument to UniversalDecoder was when it finally fixed itself, and then I backed out the scheme registration and arguments and it kept working. 

In current state the Kube RBAC is not added to scheme, at least not in main.go, and yet those resources are able to be applied. 

Will keep digging. I am setting up an OpenShift cluster to run Hive so I can test myself and save some time, I'll definitely try scheme registration in main.go, but confused as to what's going on given what I saw.